data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -95,7 +126,72 @@ private AuthenticationResult acquireAadApplicationAccessToken() throws DataServi
         return result;
     }
 
-    private Boolean isNullOrEmpty(String str) {
-        return (str == null || str.trim().isEmpty());
+    public AuthenticationResult acquireAccessTokenUsingDeviceCodeFlow() throws Exception {
+        AuthenticationContext context = null;
+        AuthenticationResult result = null;
+        ExecutorService service = null;
+        try {
+            service = Executors.newSingleThreadExecutor();
+            context = new AuthenticationContext( aadAuthorityUri, true, service);
+            Future<DeviceCode> future = context.acquireDeviceCode(CLIENT_ID, clusterUrl, null);
+            DeviceCode deviceCode = future.get();
+            System.out.println(deviceCode.getMessage());
+            if (Desktop.isDesktopSupported()) {
+                Desktop.getDesktop().browse(new URI(deviceCode.getVerificationUrl()));
+            }
+            result = waitAndAcquireTokenByDeviceCode(deviceCode, context);
+
+
+        } finally {
+            if (service != null) {
+                service.shutdown();
+            }
+        }
+        if (result == null) {
+            throw new ServiceUnavailableException("authentication result was null");
+        }
+        return result;
+    }
+
+    private AuthenticationResult waitAndAcquireTokenByDeviceCode(DeviceCode deviceCode, AuthenticationContext context)
+            throws  InterruptedException{
+        int timeout = 15 * 1000;
+        AuthenticationResult result = null;
+        while (timeout > 0){
+            try{
```

Comment:
IIUC, ExecutionException can be raised from the Future.get iff the context.acquireTokenByDeviceCode threw an exception. It that's the case, is there any value of retrying it?
Moreover, the get() should be blocking, so I am not sure there is any value of that busy wait.
Am I missing something?

Reply:
This is the type of Exception thrown when it is still pending for user authentication. So the wait is not  for future result, but for user to finish the authentication.